Add troubleshooting page (#1092)
mulesoft/docs-mule-runtime-3: modules/ROOT/pages/dataweave-troubleshoot.adoc

= Troubleshooting
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:keywords: studio, anypoint, esb, transform, transformer, format, aggregate, rename, split, filter convert, xml, json, csv, pojo, java object, metadata, dataweave, data weave, datamapper, dwl, dfl, dw, output structure, input structure, map, mapping

> 

Reviewer: I recommend that any topic at this level that contains subtopics should include some text that introduces what those subtopics are, providing a context, and tell why the audience cares (what is the user goal or advantage).

Reviewer: Is this the topic hierarchy intended?

=Troubleshooting
==Troubleshoot a Failing Script
==Exception: org.mule.weave.v2.exception.UnsupportedTypeCoercionException
=== Function Does Not Accept Null Argument
=== MIME Type Is Not Set

Try to make topic titles that are at the same hierarchical level parallel. 
I'm not sure you intend the "Exception" titled text to be a peer to the failing script text.
And are the following two topics (Function..., MIME...) subordinate to "Exception"?

== Troubleshoot a Failing Script

> One challenge to reproducing an error is having the same inputs the script had

Reviewer: ```suggestion
One challenge to reproducing an error so that you can troubleshoot it is having the same inputs the script had
```

> when it executed, especially in production environments where inputs can

Reviewer: ```suggestion
when it failed, especially in production environments where inputs can
```

change unexpectedly.

In Mule 4.2.1, DataWeave introduced an experimental feature that enables you
to dump the input context and the failing script into a folder so that you
can track the failing script along with the data that makes the script fail.

To use this feature:

. Set the following system property to enable the dump feature: `-M-Dcom.mulesoft.dw.dump_files=true`.

> . [Optional] Specify the path in which to generate the dump files: `-M-Dcom.mulesoft.dw.dump_folder=<path_to_folder>`

Reviewer: missing period
```suggestion
. [Optional] Specify the path in which to generate the dump files: `-M-Dcom.mulesoft.dw.dump_folder=<path_to_folder>`.
```

Reviewer: Consider moving the user input to its own line in each of the two steps, like this:

". Set the following property...:

`-M-Dcom.mulesoft.dw.dump_files=true`"

(No period in this case.)

+
The default directory is `/tmp`.

> == Exception: org.mule.weave.v2.exception.UnsupportedTypeCoercionException

Reviewer: Would it make sense to make this parallel to the failing script title by rewording it something like...:
```suggestion
== Troubleshooting Exceptions
```

> When a function is called with the incorrect kind of argument it throws the

Reviewer: ```suggestion
When a function is called with the incorrect kind of argument, it throws the
```

> exception, `org.mule.weave.v2.exception.UnsupportedTypeCoercionException`.

Reviewer: ```suggestion
exception `org.mule.weave.v2.exception.UnsupportedTypeCoercionException`.
```

Causes of this exception include:

* <<null_arg_not_accepted>>
* <<mimetype_unset>>

[[null_arg_not_accepted]]
=== Function Does Not Accept Null Argument

> The most common cause of this exception occurs when one of the arguments is

Reviewer: ```suggestion
The most common cause of `org.mule.weave.v2.exception.UnsupportedTypeCoercionException` occurs when one of the arguments is
```

`Null` and the function does not accept `Null` as an argument. This issue
results in an error message similar to the following:

[source,weave,linenums]
----
You called the function '++' with these arguments:
  1: Null (null)
  2: String (" A text")

> But it expects one of these combinations:

Reviewer: Is "it" the function?

Reviewer 2: that's part of the console output. We don't control it.

  (Array, Array)
  (Date, Time)
  (Date, LocalTime)
  (Date, TimeZone)
  (LocalDateTime, TimeZone)
  (LocalTime, Date)
  (LocalTime, TimeZone)
  (Object, Object)
  (String, String)
  (Time, Date)
  (TimeZone, LocalDateTime)
  (TimeZone, Date)
  (TimeZone, LocalTime)

4| payload.message ++ " A text"
   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Trace:
  at main::++ (line: 4, column: 1)
  at main::main (line: 4, column: 17)
----

One way to resolve this issue uses the `default` operator. For example,
using `payload.message default "" ++ " A text"` appends empty text when the
when the message is null.

[[mimetype_unset]]
=== MIME Type Is Not Set

When the MIME type is not set, you receive an error message similar to the
following:

[source,weave, linenums]
----
You called the function 'Value Selector' with these arguments:
  1: String ("{message: 123}")
  2: Name ("message")

> But it expects one of these combinations:

Reviewer: it = the function?

Reviewer 2: "it" is part of the console output. Docs doesn't control that.

  (Array, Name)
  (Array, String)
  (Date, Name)
  (DateTime, Name)
  (LocalDateTime, Name)
  (LocalTime, Name)
  (Object, Name)
  (Object, String)
  (Period, Name)
  (Time, Name)

4| payload.message
   ^^^^^^^^^^^^^^^
Trace:
  at main::main (line: 4, column: 1)
----

When no MIME type is set on the payload, the MIME type defaults to

> `application/java`, and the content is handled as a `String` instead of a

Reviewer: ```suggestion
`application/java`, and the content is handled as `String` instead of a
```

Reviewer 2: "a `String`" or "a `String` type" is correct, but the former is clear
